proposals/11-21-krm-fn-k8s-backend.md

```diff
@@ -0,0 +1,218 @@
+with copying labels from template will be selected. User may set the PodTemplate name with `--pod-template-name` argument and the pod
+template must exist in the same namespace where krm-function pod is going to be ran. The user may skip that parameter and by
+default the [following PodTemplate](https://github.com/kubernetes-sigs/kustomize/pull/4260/files#diff-bcc3694e89726203de4e763f2130891ec669833ab26f3856bcddde18c6514a9fR27)
+will be used. Env variables are put to that template already (TBD: maybe it's better to append env vars, rather than just replace the env vars already presented in the
+template). `--mount` flag is ignored as of now in case kubectl-backend is used, but there is a possible option to take what the set of volumes from PodTemplate
+and to append that with data from `--mount` flag - this is a potential topic for discussion.
+
+There is one more potential improvement that should be taken into account: with docker backend krm-function couldn't run
+another krm-functions. For that reason kustomize krm-function was deleted - kustomize couldn't run
+krm-functions without docker-in-docker. With kubectl backend if we could specify service account for the pod
+running krm-function we could run another krm-function initiated by the original krm-function and our
+example with kustomize-krm-function could become possible and security-aspects could be controlled with
+standard k8s-native configuration in the PodTemplate + couple of additional k8s resources.
+This is one more benefit of the approach with `--pod-template-name` flag.
+
+The proposed implementation has one more flag `--pod-start-timeout`. It was added, because depending on the cluster and the image size to pull the
+time for pod start may vary. If the user doesn't set that, the timeout is 60s by default. The previous implementation of RunFn in kyaml
+didn't work with timeouts. In contrast the updated modification that is used in `kpt` has timeouts that are applicable to the whole container execution including
+time needed to pull the image. In general, it's a good practice for automated tools to be able to set timeouts for operations. One potential topic
+for discussion may be - to create a common approach for that, e.g., to set only the overall timeout E.g. if timeout for that particular krm-functions is set
+for the whole execution, this timeout can be set in function config and we may get rid of this argument. That means we may want to stick to the approach
+that `kpt` selected.
+
+We're using the following sequence to run krm-function in the pod:
+1. If PodTemplateName is set - taking it from k8s by kubectl (the same namespace), OR using defaultPodTemplate overwise. See some mandatory fields values there, e.g. stdin,
+ that have to be set in the PodTemplate to work correctly with 'kubectl' backend.
+1. Creating pod using PodTemplate spec and applying imageUrl and set of envs
```

Comment:
How do we avoid pod name collision? e.g. If hydration pipeline X and Y run KRM function Foo at the same time.

Reply:
Good question! I forgot to cover that in the proposal.
So far I've copied/simplified [this code from k8s](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/names/generate.go#L45) that generates pod names: https://github.com/aodinokov/kustomize/blob/k8sbackend/kyaml/fn/runtime/kubectl/kubectl.go#L266
When I run kustomize I see that it generates me a new pod name each time.
Do you think we need to think about something more reliable? e.g. to use [here] https://github.com/aodinokov/kustomize/blob/k8sbackend/kyaml/fn/runtime/kubectl/kubectl.go#L273 kubectl create instead of apply and to check if kubectl returned error that the resource already existed and regenerate it?
BTW I'll do s/apply/create/ anyway..

Reply:
added part about proposed way to avoid pod-names collision. see line 137